Manticore11/bjs2 84255 system posts

Linempy opened the pull request:
Задача "Система постов"

Linempy committed:
BJS2-84255-system_posts
- написал логику сервиса для постов
- написал контроллер
- написал javaDoc на сервис и на контроллер

Linempy committed:
BJS2-84255-system_posts
- создал глобальный обработчик исключений
- написал обращение в БД по спецификации JPA для получения постов
- написал тесты для сервиса и для контроллера

RinatBeybutov commented on src/main/java/faang/school/postservice/controller/post/PostController.java:
```diff
@@ -0,0 +1,142 @@
+ * @since 26.07.2025
+ */
+@RestController
+@RequestMapping("/api/v1")
```
/api/v1 если в других контроллерах нет, то убрать. Нужно добавить /posts

RinatBeybutov commented on src/main/java/faang/school/postservice/controller/post/PostController.java:
```diff
@@ -0,0 +1,142 @@
+@RequiredArgsConstructor
+public class PostController {
+
+    private final PostServiceImpl service;
```
PostServiceImpl -> PostService если есть интерфейс

RinatBeybutov commented on src/main/java/faang/school/postservice/controller/post/PostController.java:
```diff
@@ -0,0 +1,142 @@
+     * @param userId ID пользователя
+     * @return список DTO черновиков {@link PostViewDto}
+     */
+    @GetMapping("/users/{userId}/posts/drafts")
```
прикольно придумал, но длинно, давай заменим на:
posts/{userId}

а drafts или published можно через фильтр

RinatBeybutov commented on src/main/java/faang/school/postservice/controller/post/PostController.java:
```diff
@@ -0,0 +1,142 @@
+    }
+
+    /**
+     * Получает черновики проекта, отсортированные по дате создания (новые сначала).
```
проекты = посты?

RinatBeybutov commented on src/main/java/faang/school/postservice/controller/post/PostController.java:
```diff
@@ -0,0 +1,142 @@
+     * @return DTO запрошенного поста {@link PostViewDto}
+     */
+    @GetMapping("/posts/{postId}")
+    public PostViewDto getById(@PathVariable Long postId) {
```
ResponseEntity здесь и далее

RinatBeybutov commented on src/main/java/faang/school/postservice/repository/PostRepository.java:
```diff
@@ -1,12 +1,16 @@
 import java.util.List;
 
-public interface PostRepository extends CrudRepository<Post, Long> {
+@Repository
```
убиваем

RinatBeybutov commented on src/main/java/faang/school/postservice/repository/PostRepository.java:
```diff
@@ -18,7 +22,43 @@ public interface PostRepository extends CrudRepository<Post, Long> {
 
-    @Query("SELECT p FROM Post p WHERE p.published = false AND p.deleted = false AND p.scheduledAt <= CURRENT_TIMESTAMP")
+    @Query(
+        "SELECT p FROM Post p WHERE p.published = false AND p.deleted = false AND p.scheduledAt <= CURRENT_TIMESTAMP"
```
перенести в многострочный комментарий и поправить форматирование
```
select р from Post p
left join User г on u.id = p.userId
where p.published = false
           AND p.deleted = false
group by
having
```

RinatBeybutov commented on src/main/java/faang/school/postservice/service/post/PostServiceImpl.java:
```diff
@@ -0,0 +1,162 @@
+    }
+
+    @Override
+    public PostViewDto update(Long id, PostUpdateDto dto) {
```
@Transactional

RinatBeybutov commented on src/main/java/faang/school/postservice/service/post/PostServiceImpl.java:
```diff
@@ -0,0 +1,162 @@
+        return mapper.toViewDto(post);
+    }
+
+    @Override
```
@Transactional

RinatBeybutov commented on src/main/java/faang/school/postservice/service/post/PostServiceImpl.java:
```diff
@@ -0,0 +1,162 @@
+    }
+
+    private Post findPostOrThrow(Long id) {
+        return postRepository.findById(id)
```
перенести в репозиторий

RinatBeybutov commented on src/test/java/faang/school/postservice/service/post/PostServiceImplTest.java:
```diff
@@ -0,0 +1,281 @@
+
+
+    private static Stream<Arguments> provideTestCases() {
+        return Stream.of(
```
выносим в тестовый класс PostServiceImplTestData

RinatBeybutov commented on src/test/java/faang/school/postservice/service/post/PostServiceImplTest.java:
```diff
@@ -0,0 +1,281 @@
+
+    @Test
+    @DisplayName("Выбрасывает исключение, когда ID автора и ID проекта null")
+    public void shouldThrowForbiddenException_WhenAuthorIdAndProjectIdAreNull() {
```
Нужна середина между
[myTest ... shouldThrowForbiddenException_WhenAuthorIdAndProjectIdAreNull]

testErrorWhenIdAreNull

RinatBeybutov commented on src/test/java/faang/school/postservice/service/post/PostServiceImplTest.java:
```diff
@@ -0,0 +1,281 @@
+
+    @Test
+    @DisplayName("Выбрасывает исключение, когда ID автора и ID проекта не null")
+    public void shouldThrowForbiddenException_WhenAuthorIdAndProjectIdAreNotNull() {
```
testErrorWhenBothIdNotNull

RinatBeybutov commented on src/test/java/faang/school/postservice/service/post/PostServiceImplTest.java:
```diff
@@ -0,0 +1,281 @@
+
+        verify(mapper, times(1)).toViewDto(post);
+        assertEquals(mapper.toViewDto(post), result);
+        assertEquals(currentUserId, post.getAuthorId());
```
assertThat(post)
. usingRecousiveComparsion()
. isEqualTo(expectedPost()) 
Получится использовать?

RinatBeybutov commented on src/test/java/faang/school/postservice/service/post/PostServiceImplTest.java:
```diff
@@ -0,0 +1,281 @@
+                post.isPublished(),
+                false,
+                null);
+        when(postRepository.findById(postId)).thenReturn(Optional.of(post));
```
POST_ID_1

Linempy committed:
BJS2-84255-system_posts
- в контроллере сделал центральный эндпоинт для фильтрации
- в репозитории сделал центральный метод для фильтрации
- изменил названия тестовых методов
- перенес создание данных в PostServiceImplTestData
- изменил тесты для методы findByFilters

RinatBeybutov commented on src/main/java/faang/school/postservice/controller/post/PostController.java:
```diff
@@ -89,54 +92,22 @@ public ResponseEntity<Void> softDelete(@PathVariable Long postId) {
-    @GetMapping("/users/{userId}/posts/drafts")
-    public List<PostViewDto> getByUserDraftPostsSortedByCreation(@PathVariable Long userId) {
-        return service.getByUserDraftPostsSortedByCreation(userId);
+    @PostMapping("/filter")
```
здесь getMapping, а filterDto через параметры запроса, вот пример:

```
@RestController
@RequestMapping("/api/users")
public class UserController {

    private final UserService userService;

    public UserController(UserService userService) {
        this.userService = userService;
    }

    @GetMapping
    public ResponseEntity<List<UserDto>> getFilteredUsers(@ModelAttribute UserFilterDto filter) {
        List<UserDto> users = userService.getFilteredUsers(filter);
        return ResponseEntity.ok(users);
    }
}
```

RinatBeybutov commented on src/main/java/faang/school/postservice/repository/PostSpecificationBuilder.java:
```diff
@@ -0,0 +1,63 @@
+ */
+public class PostSpecificationBuilder {
+    public static Specification<Post> buildSpecification(PostFilterDto filter) {
+        return Specification.where(byAuthorId(filter.author()))
```
можно использовать Specification.allOf(
..
)

RinatBeybutov commented on src/main/java/faang/school/postservice/repository/PostSpecificationBuilder.java:
```diff
@@ -0,0 +1,63 @@
+        }
+
+        return (root, query, cb) -> {
+            String field = author.type() == AuthorType.USER ? "authorId" : "projectId";
```
это лучше все-таки разделить на два фильтра и два разных поля

RinatBeybutov commented on src/main/java/faang/school/postservice/repository/PostSpecificationBuilder.java:
```diff
@@ -0,0 +1,63 @@
+
+        return (root, query, cb) -> {
+            String field = author.type() == AuthorType.USER ? "authorId" : "projectId";
+            return cb.equal(root.get(field), author.id());
```
чтобы совсем по взрослому выглядело, добавь генерацию наименования поля, посмотри hibernate-jpamodelgen.
У тебя эта строка будет выглядеть примерно вот так:
return cb.equal(root.get(Post_.AUTHOR_ID), author.id());

RinatBeybutov commented on src/main/java/faang/school/postservice/dto/post/PostFilterDto.java:
```diff
@@ -0,0 +1,26 @@
+public record PostFilterDto(
+    AuthorFilter author,
+    PostStatus status,
+    SortType sort,
```
сортировку можно в Pageable передавать:
```
@GetMapping
public Page<PostDto> getFilteredPosts(
        @ModelAttribute PostFilterDto filter,
        @PageableDefault(size = 20, sort = "createdAt", direction = Sort.Direction.DESC) Pageable pageable) {
    return postService.getFilteredPosts(filter, pageable);
}
```

Linempy committed:
BJS2-84255-system_posts
- исправил замечания

RinatBeybutov commented:
Красава, вот теперь функциональность фильтрации выглядит как production ready code. Ура!
